### ISSUE TITLE
Extract anonymous function on map doesn't work as expected

### ISSUE THREAD

**Issue by mynomoto** [clojure-lsp/clojure-lsp#647]
**Describe the bug**
Trying to extract an anonymous function causes extraction of the whole map form.

**To Reproduce**
1.If I have the following form (cursor represented by `|`):
```clojure
(defn map-fn
  [xs]
  (map #(a|ssoc % :a :b) xs))
```
2. Use the Extract to function code action. It will result in.
```clojure
(defn new-function [xs]
  (map #(assoc % :a :b) xs))

(defn map-fn
  [xs]
  (new-function xs))
```
3. This is not what I would expect.

**Expected behavior**
```clojure
(defn new-function [x]
  (assoc x :a :b))

(defn map-fn
  [xs]
  (map new-function xs))
```

**User details (please complete the following information):**
 - OS: ArchLinux
 - Editor: nvim
 - Version: `clojure-lsp 2021.11.16-16.52.14 clj-kondo 2021.10.20-SNAPSHOT`

**Additional context**
It should be smart about local arguments used in the anonymous function.
Discussed on slack: https://clojurians.slack.com/archives/CPABC1H61/p1637496332088400

**Comment by dkvasnicka** [comment]
The same problem occurs for forms inside the body of a `let`, where invoking the action on any of them extracts the whole `let` to a new function. But then if I go deeper into the forms and for example place the cursor inside a string then only the string is extracted - so overall very confusing behaviour.

**Comment by ericdallo** [comment]
@dkvasnicka could you give an example of the code and the expected behavior?

**Comment by mainej** [comment]
I believe this can be closed, with the addition of `promote-fn` in #922.

Starting from the original example, the first invocation of `promote-fn` (in the command menu `Promote #() to fn`) leads to:

```clojure
(defn map-fn
  [xs]
  (map (fn [element] (as|soc element :a :b)) xs))
```

The second invocation (`Promote fn to defn`), leads to:

```clojure
(defn ^:private new-function [element] (assoc element :a :b))

(defn map-fn
  [xs]
  (map new-funct|ion xs))
```

As suggested, `promote-fn` is smart about local variables used within the body of the `fn`—they are prepended to the `defn`'s argument list.

For a discussion of why this was implemented as a new refactoring, and not as a change to the behavior of `extract-function`, see the discussion starting [here](https://github.com/clojure-lsp/clojure-lsp/issues/783#issuecomment-1079478514).

@dkvasnicka if your problem is still occurring with the latest release, perhaps you could open another issue?
